### PR TITLE
Python package upgrades, preparation for the Notebook 7 upgrade

### PR DESCRIPTION
**Pull request by mkoeppe**
Cherry-picked from #36129 for easier review

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

- Depends on #36206 (merged here)
- Depends on #36263 (merged here)

### CODE REVIEW

File: build/pkgs/trove_classifiers/type [hunk 1]
```diff
@@ -0,0 +1 @@
+optional
```
Comment (orlitzky) [inline build/pkgs/trove_classifiers/type]: Hatchling is standard?

File: build/pkgs/traitlets/package-version.txt [hunk 2]
```diff
@@ -1 +1 @@
-5.5.0
+5.9.0
```
Comment (orlitzky) [inline build/pkgs/traitlets/package-version.txt:1]: Traitlets no longer needs decorator: https://github.com/ipython/traitlets/commit/7cf7f97b265f789648d78a22dd9a7d5e6e8d9d70

**Comment by orlitzky** [comment]
OK

**Comment by mkoeppe** [comment]
Thank you!

**Comment by mkoeppe** [comment]
rebased

**Comment by vbraun** [comment]
merge conflict

**Comment by mkoeppe** [comment]
Setting back to positive after merge

**Pull request merged**